libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,70 +134,148 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
                     }
+
+                    // always check the autorefresh
+                    CheckAutoRefreshToken();
                 }
             }
 
-            // We need to refresh the token, because:
-            // 1. The user requested it via the forceRefresh parameter
-            // 2. We have it, but it's expired
-            // 3. We don't have it in the cache. 
+            // if we have an oAuthTokenTask then we need to await it
+            if (oAuthTokenTask != null)
+            {
+                oAuthToken = await oAuthTokenTask;
+                tokenCache[TokenCacheKey] = oAuthToken;
+            }
 
-            OAuthResponse token = await RefreshTokenAsync().ConfigureAwait(false);
-            lock (_cacheSync)
+            return oAuthToken?.access_token;
+        }
+
+        private void CheckAutoRefreshToken()
+        {
+            // get the current autorefreshTime for this key
+            DateTime refreshTime;
+            if (autoRefreshTimes.TryGetValue(TokenCacheKey, out refreshTime))
+            {
+                // if we are past the refresh time
+                if (DateTime.UtcNow > refreshTime)
+                {
+                    // set new refresh time (this keeps only one outstanding refresh Task at a time)
+                    autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
+
+                    // background task to refresh the token
+                    // NOTE: This is not awaited, but is observed with the ContinueWith() clause.  
+                    // It is gated by the AutoRefreshTimes[] array
+                    RefreshTokenAsync()
+                        .ContinueWith(task =>
+                        {
+                            // observe the background task and put in cache when done
+                            if (task.Status == TaskStatus.RanToCompletion)
+                            {
+                                // update the cache with completed task so all new requests will get it
+                                tokenTaskCache[TokenCacheKey] = task;
+                            }
+                            else
+                            {
+                                // it failed, shorten the refresh time for another task to try again in a 30s
+                                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + TimeSpan.FromSeconds(30);
+                            }
+                        });
+                }
+            }
+        }
+
+        /// <summary>
+        /// Do not call this except from GetTokenRefreshAsync() 
+        /// </summary>
+        /// <returns></returns>
+        private Task<OAuthResponse> _getCurrentTokenTask(bool forceRefresh)
+        {
+            Task<OAuthResponse> oAuthTokenTask;
+
+            // if there is not a task or we are forcing it
+            if (forceRefresh || tokenTaskCache.TryGetValue(TokenCacheKey, out oAuthTokenTask) == false)
```

Comment:
This if statement is complex. Consider simplifying.

Reply:
yeah, I spent way too much time getting it into this state, I don't want to change it.

---
In reply to: [198677295](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198677295) [](ancestors = 198677295)